rspace/src/test/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryGenerativeSpec.scala

```diff
@@ -0,0 +1,213 @@
+    PropertyCheckConfiguration(minSuccessful = 2)
+
+  val emptyRoot: Trie               = EmptyTrie
+  val emptyRootHash: Blake2b256Hash = Trie.hash(emptyRoot)(History.codecTrie)
+
+  def lmdbEnv = {
+    val dbDir: Path           = Files.createTempDirectory("rchain-storage-test-")
+    val mapSize: Long         = 1024L * 1024L * 4096L
+    val flags: List[EnvFlags] = List(EnvFlags.MDB_NOTLS)
+    val env                   = Context.env(dbDir, mapSize, flags)
+    env
+  }
+
+  def lmdbHistoryStore = {
+    val env                  = lmdbEnv
+    val dbi: Dbi[ByteBuffer] = env.openDbi(s"history-store", MDB_CREATE)
+    new LMDBHistoryStore[Task](env, dbi)
+  }
+  def lmdbPointerBlockStore = {
+    val env                  = lmdbEnv
+    val dbi: Dbi[ByteBuffer] = env.openDbi(s"pb-store", MDB_CREATE)
+    new LMDBPointerBlockStore[Task](env, dbi)
+  }
+  def lmdbColdStore = {
+    val env                  = lmdbEnv
+    val dbi: Dbi[ByteBuffer] = env.openDbi(s"cold-store", MDB_CREATE)
+    new LMDBColdStore[Task](env, dbi)
+  }
+
+  override def repo: HistoryRepository[Task, String, Pattern, String, StringsCaptor] = {
+    val emptyHistory =
+      new History[Task](emptyRootHash, lmdbHistoryStore, lmdbPointerBlockStore)
+    val repository: HistoryRepository[Task, String, Pattern, String, StringsCaptor] =
+      HistoryRepositoryImpl.apply[Task, String, Pattern, String, StringsCaptor](
+        emptyHistory,
+        lmdbColdStore
+      )
+    repository
+  }
+}
+
+class InmemHistoryRepositoryGenerativeSpec
+    extends HistoryRepositoryGenerativeDefinition
+    with InMemoryHistoryRepositoryTestBase {
+  implicit val propertyCheckConfigParam: PropertyCheckConfiguration =
+    PropertyCheckConfiguration(minSuccessful = 10)
+
+  override def repo: HistoryRepository[Task, String, Pattern, String, StringsCaptor] = {
+    val emptyHistory =
+      new History[Task](emptyRootHash, inMemHistoryStore, inMemPointerBlockStore)
+    val repository: HistoryRepository[Task, String, Pattern, String, StringsCaptor] =
+      HistoryRepositoryImpl.apply[Task, String, Pattern, String, StringsCaptor](
+        emptyHistory,
+        inMemColdStore
+      )
+    repository
+  }
+}
+
+abstract class HistoryRepositoryGenerativeDefinition
+    extends FlatSpec
+    with Matchers
+    with OptionValues
+    with GeneratorDrivenPropertyChecks {
+
+  implicit val codecString: Codec[String]   = implicitly[Serialize[String]].toCodec
+  implicit val codecP: Codec[Pattern]       = implicitly[Serialize[Pattern]].toCodec
+  implicit val codecK: Codec[StringsCaptor] = implicitly[Serialize[StringsCaptor]].toCodec
+
+  def repo: HistoryRepository[Task, String, Pattern, String, StringsCaptor]
+
+  "HistoryRepository" should "accept all HotStoreActions" in forAll(
+    distinctListOf(arbitraryHotStoreActions)
+  ) { actions: List[HotStoreAction] =>
+    val repository = repo
+    (for {
+      _ <- actions.foldLeftM(repository) { (repo, action) =>
```

Comment:
Isn't what you're doing here actually a `traverse` disguised as `foldLeftM`?

Reply:
it is! good catch!

Reply:
oh, I was too excited. each step delivers o new repository, so it has to fold.

Reply:
But why is a new repository needed each time?

Reply:
because its stateful. the history points to a root. new data - new root.